src/linear_algebra/clifford_algebra/basic.lean

```diff
@@ -64,23 +64,116 @@ end clifford_algebra
 
 /--
 The Clifford algebra of an `R`-module `M` equipped with a quadratic_form `Q`.
+
+We make it a structure to be sure that Lean can not unfold it, and we will also use
+irreducible operations for performance reasons.
 -/
-@[derive [inhabited, ring, algebra R]]
-def clifford_algebra := ring_quot (clifford_algebra.rel Q)
+structure clifford_algebra :=
+(out : ring_quot (clifford_algebra.rel Q))
 
 namespace clifford_algebra
 
+@[irreducible] private def zero : clifford_algebra Q := ⟨0⟩
+@[irreducible] private def one : clifford_algebra Q := ⟨1⟩
+@[irreducible] private def add : clifford_algebra Q → clifford_algebra Q → clifford_algebra Q
+| ⟨a⟩ ⟨b⟩ := ⟨a + b⟩
+@[irreducible] private def neg : clifford_algebra Q → clifford_algebra Q | ⟨a⟩ := ⟨-a⟩
+@[irreducible] private def sub : clifford_algebra Q → clifford_algebra Q → clifford_algebra Q
+| ⟨a⟩ ⟨b⟩ := ⟨a - b⟩
+@[irreducible] private def mul : clifford_algebra Q → clifford_algebra Q → clifford_algebra Q
+| ⟨a⟩ ⟨b⟩ := ⟨a * b⟩
+-- next one not irreducible to avoid diamonds
+protected def smul {S : Type*} [comm_semiring S] [algebra S (tensor_algebra R M)] :
+  S → clifford_algebra Q → clifford_algebra Q
+| r ⟨a⟩ := ⟨r • a⟩
```

Comment:
To avoid diamonds it also needs to not be implemented with a pattern match.

Reply:
Done.